### PR TITLE
**Feature: ** Compact Sidenav

### LINKED ISSUE #666: Sidenav compact mode
## Summary

image

The idea is to be able to have this kind of sidenav style out of the box.

## API Proposal

```jsx
<Sidenav compact>
      <SidenavHeader label="Usecases" active>
        <SidenavItem
          label={intl.formatMessage({
            id: "sidenav.overview",
            defaultMessage: "Overview"
          })}
          icon="Home"
          to="/overview"
          active={isActive("/overview")}
        />
        <SidenavItem
          label={intl.formatMessage({
            id: "sidenav.usecases",
            defaultMessage: "Use cases"
          })}
          icon="Search"
          to="/usecases"
          active={isActive("/usecases")}
        />
        <SidenavItem
          label={intl.formatMessage({
            id: "sidenav.guides",
            defaultMessage: "Guides"
          })}
          icon="Document"
          to="/guides"
          active={isActive("/guides")}
        />
        <SidenavItem
          end
          label={intl.formatMessage({
            id: "sidenav.admin",
            defaultMessage: "Admin"
          })}
          icon="Jobs"
          to="/admin"
          active={isActive("/admin")}
        />
      </SidenavHeader>
</Sidenav>
```

## Actual issues

* The `Layout` component don't care of the `Sidenav` width for now, so we need to refactor a bit the `Layout` component to be more flexible.

## To be discussed

1. To be compatible with the normal mode, I think that we shoud keep the `SidenavHeader` layer, but hide it in compact mode.
1. The `admin` item have a `end` property, I don't know if it's the best way to do this, just a proposal. The idea is to have a way to put the item on the end of page.

### PR DESCRIPTION
**Pull request by TejasQ**
# [Go Play](http://deploy-preview-677--operational-ui.netlify.com/#!/Sidenav).

This PR adds support for a more compact sidenav component and fixes #666.

To test, check out the example. The first example _isn't_ in compact mode. Add a `compact` prop to `Sidenav` in the first example, and things should _JustWork™_. **There is a deliberate edge case in there where no `shortLabel` is provided and text overflows**.

Per the issue, there is also an `end` prop that can be added to any `SidenavHeader`, and, if it has room, it will be positioned on the bottom. By convention, this prop can and should only be applied to the _last_ `SidenavHeader` (logically) in the tree.

There is a little divergence from the issue spec here. From my perspective, `SidenavHeader` should be pinable to the bottom and then take all its children. I don't think `<SidenavItem end` makes sense because it would be strange to have a `SidenavHeader` contain _n_ items with item _n_ being at the bottom, forcing a giant height on the header.

## Related issue
#666 

## To be tested

Me
- [x] No error/warning in the console

Tester 1 (fafa)

- [x] The netlify build is working
- [x] Component looks as expected
- [x] Component works as specified in #666 
- [x] Adding a `compact` prop to a regular old `Sidenav` compacts it
  

Tester 2

- [ ] The netlify build is working
- [ ] Component looks as expected
- [ ] Component works as specified in #666 
- [ ] Adding a `compact` prop to a regular old `Sidenav` compacts it

### CODE REVIEW

**Comment by fabien0102** [comment]
About the `end` behaviour, the original idea was to add `end` to the `SidenavItem` instead of the `SidenavHeader` (and ignore the `end` props on non-compact mode).

The both works, but I think that it contiamo-ui integration will be easier with the original idea, any thought about this?

File: src/SidenavHeader/SidenavHeader.tsx [hunk 1]
```diff
@@ -32,43 +36,44 @@ export interface SidenavHeaderProps extends DefaultProps {
...
-  flexDirection: "column",
-  alignItems: "flex-start",
-  justifyContent: "center",
-  height: isCondensed ? 60 : 73,
-  overflow: "hidden",
-  padding: `0 ${theme.space.content}px`,
-  width: "100%",
-}))
+const makeContainer = (type: "link" | "block") =>
+  styled(type === "link" ? "a" : "div")<{ compact: SidenavHeaderProps["compact"]; end_: SidenavHeaderProps["end"] }>(
```
[the remark this answers is not in the document]
Comment (TejasQ) [reply src/SidenavHeader/SidenavHeader.tsx]: image

[the remark this answers is not in the document]
Comment (TejasQ) [reply src/SidenavHeader/SidenavHeader.tsx]: ¯\_(ツ)_/¯

File: src/SidenavItem/SidenavItem.tsx [hunk 2]
```diff
@@ -1,43 +1,55 @@
...
+  icon?: IconName
+  /** A label for the item when the containing sidenav is full */
   label: string
+  /** A label for the item when the containing sidenav is compact */
+  shortLabel?: string
+  compact?: SidenavHeaderProps["compact"]
 }
 
-const size: number = 36
+const getSize = (compact = false) => (compact ? 30 : 18)
```
[the remark this answers is not in the document]
Comment (TejasQ) [reply src/SidenavItem/SidenavItem.tsx]: Good catch!

**Comment by TejasQ** [comment]
About the `end` behavior, it's a little tricky because the use case we have is that a `SidenavItem` should be pinnable to the bottom, _not_ a `SidenavHeader`. I've just implemented this in the most recent commit.

For integration with Contiamo UI, this is not currently specified. I think embeddable apps should not have to concern themselves with the entirety of Contiamo UI's sidenavs and as such, should not export `end` SidenavItems.

I guess we'll figure that out as we go in the coming weeks.

File: src/Sidenav/README.md [hunk 3]
```diff
@@ -31,27 +33,28 @@ class StatefulSidenav extends React.Component {
             active={this.state.activeHeaders.includes(1)}
             onToggle={this.toggle.bind(this, 1)}
           >
-            <SidenavItem label="The First Prize" icon="Settings" />
-            <SidenavItem label="The Second Prize" icon="Settings" />
-            <SidenavItem label="The Third Prize" icon="Settings" />
+            <SidenavItem label="The First Prize" shortLabel="First" icon="Add" />
```
Comment (peterszerzo) [inline src/Sidenav/README.md]: Use `compactLabel` to make it more clear that this label will be used in `condensed` mode.

File: src/Sidenav/Sidenav.tsx [hunk 4]
```diff
@@ -1,33 +1,48 @@
...
 })
 
-const Sidenav: React.SFC<SidenavProps> = ({ children, ...props }) => <Container {...props}>{children}</Container>
+const Sidenav: React.SFC<SidenavProps> = ({ children, ...props }) => (
+  <Container {...props}>
+    {React.Children.map(children, child => {
+      // see line 8
+      const dudeMyChildIsAnElementJustLetMeUseItUgh = child as React.ReactElement<any>
+      return {
+        ...dudeMyChildIsAnElementJustLetMeUseItUgh,
```
Comment (peterszerzo) [inline src/Sidenav/Sidenav.tsx]: `unsafelyTypedReactChild` maybe? Otherwise it takes a long time to read and still not clear that it's a typescript workaround.

**Comment by TejasQ** [comment]
[Support bottom item in sidebar](https://trello.com/c/2D4oz7Jp/384-support-bottom-item-in-sidebar)

File: src/SidenavHeader/SidenavHeader.tsx [hunk 5]
```diff
@@ -79,7 +79,8 @@ const LabelText = styled("div")<{ isActive: boolean; compact: SidenavHeaderProps
 `
 
 const ItemsContainer = styled("div")({
-  animation: `${floatIn} .15s forwards ease`,
+  /** @todo add this animation when we move to a JSON-style API for SidenavHeaders */
```
Comment (peterszerzo) [inline src/SidenavHeader/SidenavHeader.tsx]: Can we just add here that this had to be removed because the animation reran every single time the component updated, because of the element cloning in the `Sidenav`'s implementation?
